### PR TITLE
WIP: dracut: init at 049

### LINKED ISSUE #77858: Packaging dracut
**Project description**

dracut enhances the initramfs/initrd. Primarily I was looking for it userspace tools like `lsinitrd`.

Official github repo: https://github.com/dracutdevs/dracut (it might be a kernel mirror, or the kernel might be a mirror of the github, not sure).

Anyway my progress in trying to package this was:

1. Create shell.nix

```
{ pkgs ? import <nixpkgs> {} }:

with pkgs;
stdenv.mkDerivation rec {
  name = "dracut";
  src = null;
  nativeBuildInputs = [ pkg-config ];
  shellHook = ''
    echo 'Entering ${name}'
  '';
}
```

2. Then run `patchShebangs ./configure` (it has /bin/bash)

3. Then run ./configure

The result is:

```
Package libkmod was not found in the pkg-config search path.
Perhaps you should add the directory containing `libkmod.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libkmod' found
dracut needs pkg-config and libkmod >= 23.
```

I cannot find `libkmod` and I don't know where to even get that.

### PR DESCRIPTION
**Pull request by CMCDragonkai**
## Motivation for this change

Fixes #77858

This derivation still has some problems. In particular `realpath` is not found in the dracut command. And also the `dracutbasedir` has shell scripts there that have not been properly wrapped as well. And also I need to wrap the bin scripts with a PATH to itself, but not sure if that means I need to add `$out` to the `lib.makeBinPath`.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by CMCDragonkai** [comment]
Also is running `wrapProgram` like that recommended?

**Comment by CMCDragonkai** [comment]
If somebody knows more how to package this better, please recommend.

File: pkgs/os-specific/linux/dracut/default.nix [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+, gzip
+, bzip2
+, lz4
+, lzop
+, zstd
+, xz
+}:
+
+stdenv.mkDerivation rec {
+  name = "dracut";
```
Comment (ajs124) [inline pkgs/os-specific/linux/dracut/default.nix]: why name and not pname?

File: pkgs/os-specific/linux/dracut/default.nix [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  version = "049";
+  src = fetchurl {
```
Comment (ajs124) [inline pkgs/os-specific/linux/dracut/default.nix]: Is there a reason why you didn't use fetchFromGitHub?

**Review by ajs124 — CHANGES_REQUESTED** [review]
No idea if `wrapProgram` is intended to be used like that.

**Comment by CMCDragonkai** [comment]
Does fetchFromGithub get releases? I forgot.

Any alternative to setting the PATH for all the scripts?

On 19 January 2020 08:29:05 GMT+11:00, ajs124 <notifications@github.com> wrote:
>ajs124 requested changes on this pull request.
>
>No idea if `wrapProgram` is intended to be used like that.
>
>> +, utillinux
>+, gnused
>+, gnugrep
>+, squashfsTools
>+, cpio
>+, binutils
>+, gzip
>+, bzip2
>+, lz4
>+, lzop
>+, zstd
>+, xz
>+}:
>+
>+stdenv.mkDerivation rec {
>+  name = "dracut";
>
>why name and not pname?
>
>> +, gnugrep
>+, squashfsTools
>+, cpio
>+, binutils
>+, gzip
>+, bzip2
>+, lz4
>+, lzop
>+, zstd
>+, xz
>+}:
>+
>+stdenv.mkDerivation rec {
>+  name = "dracut";
>+  version = "049";
>+  src = fetchurl {
>
>Is there a reason why you didn't use fetchFromGitHub?
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/77868#pullrequestreview-344963396

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: pkgs/os-specific/linux/dracut/default.nix (hunk 1 above)
Comment (CMCDragonkai) [reply pkgs/os-specific/linux/dracut/default.nix]: `pname` is just name right and then combined with `version` right?

File: pkgs/os-specific/linux/dracut/default.nix (hunk 2 above)
Comment (CMCDragonkai) [reply pkgs/os-specific/linux/dracut/default.nix]: Does that fetch the releases?

File: pkgs/os-specific/linux/dracut/default.nix [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  preConfigure = ''
+    patchShebangs ./configure
+  '';
+  postPatch = ''
+    substituteInPlace dracut.sh \
+      --replace 'dracutbasedir=/usr/lib/dracut' "dracutbasedir=$out/lib/dracut"
+    substituteInPlace lsinitrd.sh \
+      --replace 'dracutbasedir=/usr/lib/dracut' "dracutbasedir=$out/lib/dracut"
+  '';
+  postFixup = ''
```
Comment (doronbehar) [inline pkgs/os-specific/linux/dracut/default.nix]: > Any alternative to setting the PATH for all the scripts?

I think @CMCDragonkai you can do something like this (not tested):

```suggestion
  # put here all the packages you need - wrap them all
  wrapPATH = lib.makeBinPath [ ... ];
  postFixup = ''
    wrapProgram $out/bin/dracut --set PATH ${wrapPATH}
  '';
```

BUT, TBH I'm a bit conflicted whether **all** of those packages are needed as they are expected to be available in almost every environment. Personally, I'd drop at least:

- `gzip`
- `bzip2`
- `xz`
- `gnugrep`
- `gnused`

BTW I assume `makeBinPath` uses the order of the packages it gets in order to give precedence in `$PATH`. Hence,  If upstream requires explicitly `gnused` instead whatever's in `coreutils`, you should put `gnused` before `coreutils`.

File: pkgs/os-specific/linux/dracut/default.nix [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    ]}
+    wrapProgram $out/bin/mkinitrd --set PATH ${lib.makeBinPath [
+      coreutils
+      gnused
+      gnugrep
+    ]}
+  '';
+  meta = with lib; {
+    description = "dracut is an event driven initramfs infrastructure";
+    homepage = https://dracut.wiki.kernel.org;
```
Comment (doronbehar) [inline pkgs/os-specific/linux/dracut/default.nix]:
```suggestion
    homepage = "https://dracut.wiki.kernel.org";
```

**Review by doronbehar — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/linux/dracut/default.nix (hunk 3 above)
Comment (CMCDragonkai) [reply pkgs/os-specific/linux/dracut/default.nix:53]: I think the derivation should be as pure as possible to ensure that if I package `dracut` in a docker container it should have everything.

**Comment by lilyinstarlight** [comment]
I've continued this PR in #210075

**Pull request closed without merging**